Improved documentation on DC/OS installation with a custom CA certificate

ahairape opened the pull request:
## Description
https://jira.mesosphere.com/browse/DCOS-19638

## Urgency
- [ ] Blocker <!-- Ping @sascala, @stbof, or @pavisandhu for review -->
- [ X ] High
- [ ] Medium

This is detailed documentation on installing DC/OS with a custom CA certificate. (improvement on the existing doc)

mhrabovcin commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,317 @@ excerpt:
 
-Custom CA certificate support is enabled by three install time [configuration parameters](/1.10/installing/ent/custom/configuration/configuration-parameters/#ca-certificate-path-enterprise):
+# The structure of this document page
+To facilitate the the reading of this page we start out by providing a Glossary for general definition of terms, followed by an in-depth configuration parameter reference. An installation walkthrough is provided in section Installing DC/OS Enterprise with a custom CA certificate. Section Example use cases then provides example file contents for the custom CA certificate configuration files for th ...
```
`To facilitate the the` - remove one `the`.

mhrabovcin replied:
`a Glossary`, should be `a glossary` (no caps)

mhrabovcin replied:
Ca we make `Installing DC/OS Enterprise with a custom CA certificate` link to the section?

mhrabovcin commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,317 @@ excerpt:
+## Manually placing the private key associated with the custom CA certificate onto the master nodes
+
+For security reasons, the installer will not copy the private key from the bootstrap node to the master nodes. 
+The private key associated with the custom CA certificate has to be distributed manually to every DC/OS master node before starting the installation. The filesystem path for the private key file has to be `/var/lib/dcos/pki/tls/CA/private/custom_ca.key`. The directory `/var/lib/dcos/pki/tls/CA/private` should be created manually with the following  command before putting the file `custom_ca.key`  ...
```
I'd make `before starting the installation` bold :-)

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,317 @@ excerpt:
-- Ensuring that all X.509 certificates used within the DC/OS cluster (for both signing and encrypting) derive from your organization's X.509 certification hierarchy.
-- Controlling security properties of the key pair (such as type and strength) used for signing DC/OS component certificates.
-- Ensuring that all DC/OS components (including Admin Router) present browser-trusted certificates.
+- Ensuring that all X.509 certificates used within the DC/OS cluster (for both signing and encrypting) derive from your organization’s X.509 certification hierarchy
```
Let's use a normal apostrophe here. `'`.

jgehrcke replied:
I also think the full stop should stay.

jgehrcke commented:
Reviewers read it here: https://github.com/mesosphere/dcos-docs-site/blob/75a4bedb78fd40568885476029842abfc21fb156/pages/1.10/security/ent/tls-ssl/ca-custom/index.md

mhrabovcin commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,321 @@ excerpt:
+- The custom CA private key has been securely placed on all DC/OS master nodes and is sufficiently protected (refer to [this section](#manually-placing-the-private-key-associated-with-the-custom-ca-certificate-onto-the-master-nodes) for more details). Example (command issued on one of the DC/OS master nodes):
+
+```bash
+stat /var/lib/dcos/pki/tls/CA/private/dcos-ca-certificate-key.key
````
Putting a private key onto this location would work. A private key has to be always on ``/var/lib/dcos/pki/tls/CA/private/custom_ca.key` path. I think we should update this section to not confuse users.

jgehrcke replied:
... would *not* work. :)

mhrabovcin commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,321 @@ excerpt:
+depth=0 C = US, ST = CA, L = San Francisco, O = "Mesosphere, Inc.", CN = AdminRouter on 172.31.12.45
+verify return:1
+ 0 s:/C=US/ST=CA/L=San Francisco/O=Mesosphere, Inc./CN=AdminRouter on 172.31.12.45
+   i:/DC=io/DC=integration-test/C=DE/ST=Utopia/O=DC/OS/OU=Brogrammer Unit/CN=Integration Test Intermediate CA 02
```
`Brogrammer Unit` made it our public documentation 🎉 😆

jongiddy replied:
I'm not sure that's a good thing. The idea of a Brogrammer Unit might be seen to reinforce some unfortunate stereotypes. I'm sure we could have an equally amusing name that's less open to misinterpretation.

jgehrcke replied:
Thanks for raising that point, Jon. I agree. Internally it can be considered amusing, but to the public this is an unnecessary attack surface.

As one of the last polishing steps this should be replaced with a different word. @jongiddy do you volunteer for finding the equally amusing name? :)

@ahairape please own that point.

Please also make sure that we're not publishing the corresponding files.

jongiddy replied:
Schlaraffenland? Given that it exists in the state of Utopia

ahairape replied:
I think we could may be use Schlaraffenland as a value in the **C** or **ST** fields of the certificate subject, but it does not make sense to me in the **OU** field. I'll change it to a prosaic _Programmer Unit_ for now, we can change it to something more amusing later.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,322 @@ excerpt:
+
+```bash 
+cd $DCOS_INSTALL_DIR
+cat genconf/dcos-ca-certificate-chain.crt dcos-ca-certificate-chain.crt | awk -v cmd="openssl x509 -issuer -subject -noout && echo" '/-----BEGIN/ { c = $0; next } c { c = c "\n" $0 } /-----END/ { print c|cmd; close(cmd); c = 0 }'
````
This shows a file path twice here: `cat genconf/dcos-ca-certificate-chain.crt dcos-ca-certificate-chain.crt`.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+## ca\_certificate\_chain\_path
+Path to a file within the `genconf/` directory containing the complete CA certification chain required for end-entity certificate verification, in the OpenSSL PEM format. For example: genconf/CA_cert_chain.pem.
+Has to be left undefined if `ca_certificate_path` points to a root CA certificate.
+Required if ca_certificate_path is specified and if the custom CA certificate is an *intermediate CA certificate*. 
```
Backticks around `ca_certificate_path` are missing.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+In order to do that you need first to obtain the DC/OS CA bundle of the deployed cluster. [This page](/1.10/security/ent/tls-ssl/get-cert/) shows how you can do that.
+
+Provided you have obtained the DC/OS bundle and stored it in a file named `dcos-ca.crt`, issue the following command in the directory containing the `dcos-ca.crt` file in order to check that the Admin Router on a master node uses a certificate signed with the custom CA certificate:
```
Let's please say "DC/OS CA bundle" instead of "DC/OS bundle".

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
 
-- Only custom CA certificates that have an associated RSA-type key pair are supported.
-- You can use a custom CA certificate only with a new 1.10 install.
+Your custom CA certificate in the PEM format, which will be used to issue certificates for DC/OS components such as Admin Router. The custom CA certificate is either an intermediate CA certificate (issued by another CA) or a root CA certificate (self-signed by the custom CA).
```
Let us please make sure that the indentation level in the rendered page makes it easy to associate terms with the description, visually. I think I would want us to make terms in bold, and to then continue describing them _in the same bullet_. Rendered example:

- **Custom CA certificate**: Your custom CA certificate in the PEM format, which will be used to issue certificates for DC/OS components such as Admin Router. The custom CA certificate is either an intermediate CA certificate (issued by another CA) or a root CA certificate (self-signed by the custom CA).

- **Private key associated with the custom CA certificate**: ...

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+# Requirements
+
+- Advanced installation method
+In order to install DC/OS Enterprise with a custom CA certificate you have to use the [advanced DC/OS installation method](/1.10/installing/custom/advanced/). Other installation methods are not supported.
```
This is broken when rendered:
```
Advanced installation method In order to install DC/OS Enterprise with a custom CA certificate you have to use the advanced DC/OS installation method. 
```

Reviewers and @ahairape please look at the rendered result.

jongiddy replied:
I suggest starting the Requirement section with:

In order to install DC/OS Enterprise with a custom CA certificate you will need:
- to use the [advanced DC/OS installation method](/1.10/installing/custom/advanced/). Other installation methods are not supported.
- a file containing... (as suggested by @jgehrcke below)

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+- Private key associated with the custom CA certificate
+- Certificate chain associated with the custom CA certificate
+
+The files containing the custom CA certificate, the associated private key and the certificate chain has to be placed in the predefined locations where the installer can find them during the installation. (see the next sections for details)
```
The files ... have to.

jgehrcke replied:
I think we should leave this entire sentence out.

The next sentence effectively duplicates the information.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+- Advanced installation method
+In order to install DC/OS Enterprise with a custom CA certificate you have to use the [advanced DC/OS installation method](/1.10/installing/custom/advanced/). Other installation methods are not supported.
+
+- Custom CA certificate
```
Can we say "a file containing the custom CA certificate" instead?

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+In order to install DC/OS Enterprise with a custom CA certificate you have to use the [advanced DC/OS installation method](/1.10/installing/custom/advanced/). Other installation methods are not supported.
+
+- Custom CA certificate
+- Private key associated with the custom CA certificate
```
Can we say "a file containing the private ....." instead?

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+- Private key associated with the custom CA certificate
+- Certificate chain associated with the custom CA certificate
+
+The files containing the custom CA certificate, the associated private key and the certificate chain has to be placed in the predefined locations where the installer can find them during the installation. (see the next sections for details)
```
Here we can then say "The files have to be placed ..." and can leave out the duplicated information "containing the custom CA certificate, the associated private key and the certificate chain"

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+The set of the configuration parameters that governs the specific aspects of the installation procedure. The configuration is stored in the configuration file.
+
+- DC/OS Configuration file
```
"DC/OS configuration file"

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+- DC/OS Configuration file
+
+The file which contains the DC/OS configuration parameters, in particular, the  filesystem relative paths to the custom CA certificate, associated private key and the certificate chain. 
```
Let's either leave out ", in particular, the  filesystem relative paths to the custom CA certificate, associated private key and the certificate chain." or improve the grammar of this sentence.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+The file which contains the DC/OS configuration parameters, in particular, the  filesystem relative paths to the custom CA certificate, associated private key and the certificate chain. 
+
+The configuration file is normally called `config.yaml` and has to be present in the `genconf/` directory on the bootstrap node during the installation. It is used by the DC/OS installer to create a custom DC/OS build.
```
Let's continue to call it "DC/OS configuration file".

jgehrcke replied:
"to create a custom DC/OS build." is not correct. Let's leave this out.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+- Installation directory
+
+The directory on the bootstrap node where the DC/OS installer resides. It contains the `genconf/` subdirectory where the configuration file, the custom CA certificate, the associated private key and the certificate chains are put. It is denoted with `$DCOS_INSTALL_DIR` in this document. This is the current working directory when executing `dcos_generate_config.ee.sh`.
```
Let's continue to call it "DC/OS configuration file", not just "configuration file". Consistency is super helpful for a reader new to the topic.

jgehrcke replied:
There is no such thing as "the certificate chains". Let's please use terms consistently.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+- Configuration
+
+The set of the configuration parameters that governs the specific aspects of the installation procedure. The configuration is stored in the configuration file.
```
"DC/OS configuration file"

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+- Certificate chain associated with the custom CA certificate
+
+The complete CA certification chain required for end-entity certificate verification. It has to include the certificates of all parent CAs of the intermediate custom CA up to and including the self-signed root CA certificate. Empty, if the custom CA certificate is a self-signed root CA certificate.
```
We can leave out "self-signed", I think "root" is precise enough.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+
+## Manually placing the the custom CA certificate, the associated private key and the certificate chain onto the bootstrap node
+The custom CA certificate, the associated private key and the certificate chain files have to be put in the `genconf/` directory on the bootstrap node. The paths to these files have to be specified in the configuration file.
```
Below you write

```+- On the bootstrap node, the files carrying custom CA certificate, the associated private key and, optionally, the CA certificate chain have been placed into the `$DCOS_INSTALL_DIR/genconf/` directory (see the [section](#glossary-of-the-terms-used-in-this-documentation) above for more detailed description), Example (commands executed on the bootstrap node):
```

I think this section here can disappear.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+## Manually placing the the custom CA certificate, the associated private key and the certificate chain onto the bootstrap node
+The custom CA certificate, the associated private key and the certificate chain files have to be put in the `genconf/` directory on the bootstrap node. The paths to these files have to be specified in the configuration file.
+
```
Not really happy with this "action" being a "requirement". I think this should also be combined with 

```
+```bash
 +cd $DCOS_INSTALL_DIR
 +ls genconf/
 +```
 +```bash 
 +dcos-ca-certificate.crt
 +dcos-ca-certificate-key.key
 +dcos-ca-certificate-chain.crt
 +```
```

below.

jgehrcke replied:
When they read

> The custom CA certificate, the associated private key and the certificate chain files have to be put in the `genconf/` directory on the bootstrap node.

they will be a little confused if this is now something they should do right now, or if this is going to be explained in more detail below.

I think this should be a more specific call to action with the example that's actually given further below, and the vague "The paths to these files have to be specified in the configuration file." should go away (is a more specific call to action further below).

(this is the stuff I only notice when looking at a beautifully rendered document, and that I didn't really watch out before).

ahairape replied:
ok, I have moved the 'ls' command from the "Installing DC/OS Enterprise with a custom CA certificate", "Starting point" subsection to this section. (combined it with the "The custom CA certificate, the associated private key and the certificate chain files have to be put in the `genconf/` directory on the bootstrap node." sentence)
If this is not what you meant please leave a comment.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+The filesystem paths to the custom CA certificate, associated private key and certificate chain files in the `genconf` directory on the bootstrap node have to be specified in the configuration file using, respectively, the `ca_certificate_path`, `ca_certificate_key_path` and  `ca_certificate_chain_path` parameters.
+The paths to the custom certificate, the associated private key and the certificate chain must be relative paths and must not contain the `genconf` prefix.
+
+[This section](#example-use-cases) shows how to set these configuration parameters.
```
The current link looks like it might link to a separate page. Indicate to the reader that they can keep reading linearly, and will get to this information:

The [Example Use Cases] (#example-use-cases) section below shows...

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+## Starting point
+Based on the requirements described above, this is the starting point for the installation:
+
+- The installation of DC/OS Enterprise via the Advanced Installer has been prepared according to the corresponding [documentation](/1.10/installing/ent/custom/advanced/).
```
It appears that the following steps must be done between the Configure and Install steps of the linked page. It might help to be more specific about where to stop in the Advanced installation.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,322 @@ excerpt:
+ca_certificate_chain_path: genconf/dcos-ca-certificate-chain.crt
+[...]
+```
+Note that `ca_certificate_chain_path` is an optional parameter when setting up DC/OS Enterprise with a custom CA certificate when the latter is a self-signed root certificate.
````
Describing this as an "optional parameter" suggests that it is not important whether it is present or not. However, other sections suggest that it should definitely not be present in this case. Make it more direct, e.g.:

Note that `ca_certificate_chain_path` must not be present when setting up DC/OS Enterprise with a self-signed root certificate as the custom CA certificate.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
 
-Custom CA certificate support is enabled by three install time [configuration parameters](/1.10/installing/ent/custom/configuration/configuration-parameters/#ca-certificate-path-enterprise):
+# The structure of this document page
+To facilitate the reading of this page we start out by providing a glossary for general definition of terms, followed by an in-depth configuration parameter reference. An installation walkthrough is provided in section [Installing DC/OS Enterprise with a custom CA certificate](#installing-dcos-enterprise-with-a-custom-ca-certificate). Section [Example use cases](#example-use-cases) then provides  ...
```
I like this overview. It's a relatively complex page, so this gives the reader an idea of what they will find here.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
-- `ca_certificate_chain_path`
+# What is supported and what is not
+- Only custom CA certificates that have an associated RSA-type key pair are supported. Other types of certificates, such as those using ECC-type key pair, are currently not supported. Support for ECC-type key pairs will be added in the future.
+- You have to use a custom CA certificate only with a fresh installation of DC/OS Enterprise 1.10 or higher. Older versions of DC/OS are not supported.
```
Does the use fo the word "fresh" mean that it is not possible to perform these actions during an upgrade? Perhaps add this explicitly:

- Custom CA certificates are only supported for a fresh installation of DC/OS Enterprise 1.10 or higher. Older versions of DC/OS are not supported, and it is not possible to add a custom CA certificate during an upgrade.

ahairape replied:
@jgehrcke @mhrabovcin could you please confirm that using custom CA is not supported by the DC/OS upgrade? I haven't tried updating DC/OS so far. 
The existing doc does not mention the upgrade explicitly: https://docs.mesosphere.com/1.10/security/ent/tls-ssl/ca-custom/

jgehrcke replied:
@ahairape yes, this is the case. You can find this information in the design document. Please follow @jongiddy's suggestion, I like it very much.

ahairape commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,322 @@ excerpt:
+
+- Certificate chain associated with the custom CA certificate
+
+The complete CA certification chain required for end-entity certificate verification. It has to include the certificates of all parent CAs of the intermediate custom CA up to and including the self-signed root CA certificate. Empty, if the custom CA certificate is a self-signed root CA certificate.
```
There was, indeed, a contradiction in the definition of the chain that had to "contain at least one certificate" and could be empty. I have modified the Configuration parameter reference to mention explicitly that it's the CA chain for a non-root CA certificate that is required to include  at least one certificate.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,306 @@ excerpt:
+
+```bash
+cd $DCOS_INSTALL_DIR/config
+scp custom_ca.key centos@W.X.Y.Z:/var/lib/dcos/pki/tls/CA/private
````
To make the examples follow the flow, please make this command use the same filename as the bootstrap node above, i.e.:
```
scp dcos-ca-certificate-key.key centos@W.X.Y.Z:/var/lib/dcos/pki/tls/CA/private/custom_ca.key
```

If the installer copies the key from the bootstrap node, then this command will work properly, and even if they copy from another machine, they probably have used the same filename as for the bootstrap node.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,306 @@ excerpt:
+- to use the **Advanced installation method:**  [advanced DC/OS installation method](/1.10/installing/custom/advanced/). Other installation methods are not supported.
+- **A file containing the custom CA certificate**
+- **A file containing the private key associated with the custom CA certificate**
+- If the CA is **not** a root CA, **a file containing the certificate chain associated with the custom CA certificate** 
```
To avoid ambiguity, please change this to "If the custom CA certificate is **not** a self-signed root CA certificate, ..."

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,306 @@ excerpt:
+ca_certificate_chain_path: genconf/dcos-ca-certificate-chain.crt
+[...]
+```
+Note that `ca_certificate_chain_path` must not be present when setting up DC/OS Enterprise with a root certificate as the custom CA certificate.
````
Again, to avoid ambiguity, "...with a self-signed root CA certificate..."

ahairape replied:
The glossary actually mentions that the root CA certificate is one which is self-signed, so it should not be ambiguous. But in principle it does not hurt to mention it once again. @jgehrcke  asked in another comment to change "self-signed root CA certificate" to "root CA certificate", that's why I removed it through the doc entirely.

jongiddy replied:
OK, fine to leave it as is then.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,306 @@ excerpt:
+## Manually placing the private key associated with the custom CA certificate onto the master nodes
+
+For security reasons, the installer will not copy the private key from the bootstrap node to the master nodes. 
+The private key associated with the custom CA certificate must be distributed manually to every DC/OS master node **before starting the installation**. The filesystem path for the private key file must be `/var/lib/dcos/pki/tls/CA/private/custom_ca.key`. The directory `/var/lib/dcos/pki/tls/CA/private` should be created manually with the following  command before putting the file `custom_ca.key`  ...
```
In the last sentence, I think "should" should be replaced with "can". This provides a simple way to perform the steps, but makes it clear that the installer can choose to achieve the same effect in other ways (e.g. using a configuration management tool).

Note, with Markdown, it is helpful when reviewing and diff'ing if each sentence is placed on a separate line. As long as there are no blank lines, the sentences will remain in the same paragraph when rendered.

jongiddy commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,306 @@ excerpt:
+ mkdir -p /var/lib/dcos/pki/tls/CA/private
+```
+
+Furthermore, the file must be owned by the root Unix user and to have 0600 permissions set.
````
Since the file has not been copied here yet, it looks like you mean that the `private` directory should be owned by `root` and have 0700 permissions (allowing read/write/traversal. This provides an extra layer of security in case the key file does not immediately have the correct permissions while being transferred.

It might be simpler to add "As the root user" before the `mkdir` command and `chmod 700 /var/lib/dcos/pki/tls/CA/private` after the `mkdir`.

In any case, if this line remains, "to have" -> "have" or "must have".

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
````diff
@@ -8,26 +8,306 @@ excerpt:
+
+```bash 
+cd $DCOS_INSTALL_DIR
+cat genconf/dcos-ca-certificate-chain.crt | awk -v cmd="openssl x509 -issuer -subject -noout && echo" '/-----BEGIN/ { c = $0; next } c { c = c "\n" $0 } /-----END/ { print c|cmd; close(cmd); c = 0 }'
````
Spent some 20 minutes on that on Wednesday. Did not find something significantly better.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+- **Private key associated with the custom CA certificate:** The private key in the PKCS#8 format associated with the custom CA certificate.
 
-# Procedure
+- **Certificate chain associated with the custom CA certificate:** The complete CA certification chain required for end-entity certificate verification. It must include the certificates of all parent CAs of the intermediate custom CA up to and including the root CA certificate. If the custom CA certificate is a root CA certificate, the chain must be an empty.
```
>  the chain must be an empty.

Broken sentence.

What do you think about "If the custom CA certificate is a root CA certificate the chain is empty (in which case the parameter `ca_certificate_chain_path` described below must not be provided)"?

ahairape replied:
fixed

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+Furthermore, the file containing the private key `custom_ca.key` corresponding to the custom CA certificate must be owned by the root Unix user and have 0600 permissions set.
+
+If you copy private key file over the network onto the master nodes, the network channel must be adequately protected.
+An example of copying the CA private key is given below. The commands are executed on the bootstrap node. The `W.X.Y.Z` denotes the IP address of a master node below.
```
> The `W.X.Y.Z` denotes the IP address of a master node below.

Should that not be

>  The `W.X.Y.Z` below denotes the IP address of a master node.

?

ahairape replied:
yes, corrected

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+## Specifying the locations of the custom CA certificate, the associated private key and the certificate chain files in the DC/OS configuration file
+
+The filesystem paths to the custom CA certificate, associated private key and certificate chain files in the `$DCOS_INSTALL_DIR/genconf/` directory on the bootstrap node must be specified in the DC/OS configuration file using, respectively, the `ca_certificate_path`, `ca_certificate_key_path` and  `ca_certificate_chain_path` parameters.
+The paths to the custom certificate, the associated private key and the certificate chain must be relative paths. (without the `$DCOS_INSTALL_DIR` prefix)
```
Let us please write "custom CA certificate" instead of "custom certificate", for consistency.

jgehrcke replied:
Instead of correcting that, let us simplify.

Instead of 

> The paths to the custom certificate, the associated private key and the certificate chain must be relative paths. (without the `$DCOS_INSTALL_DIR` prefix)

let us write

> The file paths must be relative to `$DCOS_INSTALL_DIR`

In the Google Doc in the Glossary we added `This is the current working directory when executing dcos_generate_config.ee.sh.` to the explanation of "Installation directory". Where and why did this get lost? Was this intentional?

ahairape replied:
@jgehrcke I rephrased the explanation `This is the current working directory when executing dcos_generate_config.ee.sh.` in the **Installation** section. I moved it into that section in response to Jon's remark that the glossary terms normally define concepts and do not specify technical details. Do you want me to move it back to the glossary in its original form?
To me this important information looks equally ok in both sections as long as it is not removed from the document.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+
+# Configuration parameter reference
+## ca\_certificate\_path
+Path (realtive to the `$DCOS_INSTALL_DIR`) to a file containing a single X.509 CA certificate in the OpenSSL PEM format. For example: `genconf/dcos-ca-certificate.crt`. It is either a *root CA certificate* (“self-signed”) or an *intermediate CA certificate* (“cross-certificate”) signed by some other certificate authority.
```
> realtive

please run a spellchecker!

ahairape replied:
done

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+The public key associated with the custom CA certificate must be of type RSA.
+
+## ca\_certificate\_key\_path
+Path (realtive to the `$DCOS_INSTALL_DIR`) to a file containing the private key corresponding to the custom CA certificate, encoded in the OpenSSL (PKCS#8) PEM format. For example: `genconf/CA_cert.key`.
```
> realtive

(:

ahairape replied:
fixed

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+
+- On the bootstrap node, the files carrying custom CA certificate, the associated private key and, optionally, the CA certificate chain have been placed into the `$DCOS_INSTALL_DIR/genconf/` directory. (see the [section](#manually-placing-the-custom-ca-certificate-the-associated-private-key-and-the-certificate-chain-onto-the-bootstrap-node) above for more detailed description)
+
+- The custom CA private key has been securely placed on all DC/OS master nodes and is sufficiently protected (refer to [this section](#manually-placing-the-private-key-associated-with-the-custom-ca-certificate-onto-the-master-nodes) for more details). Example (command issued on one of the DC/OS master nodes):
```
"custom CA private key" is a term that was not used in this document so far. Please use terms consistently.

jgehrcke replied:
Please remove "and is sufficiently protected". It does not contain useful information.

ahairape replied:
corrected

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+
+## Installation
+Proceed with the installation as described in the [documentation of the Advanced Installer](/1.10/installing/ent/custom/advanced/#install-dcos).
+Note that the current working directory when executing `dcos_generate_config.ee.sh` must be the `$DCOS_INSTALL_DIR` directory.
```
Well, the current directory when executing dcos_generate_config.ee.sh _defines_ $DCOS_INSTALL_DIR used here.

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+Note that the current working directory when executing `dcos_generate_config.ee.sh` must be the `$DCOS_INSTALL_DIR` directory.
+
+## Verify installation
+One method of verifying that the DC/OS was installed properly with the custom CA certificate is to initiate a TLS connection to the Admin Router which, after installation, will have a certificate signed by the custom CA. 
```
> that the DC/OS 

"the DC/OS" must be changed. "the DC/OS Enterprise cluster"?

jgehrcke replied:
We also do not say "the Admin Router". Use "Admin Router" instead, to be consistent with the larger fraction of the rest of the docs.

jgehrcke replied:
Can we say "will present a certificate" instead of "will have a certificate"?

ahairape replied:
yes, corrected

jgehrcke commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+## Verify installation
+One method of verifying that the DC/OS was installed properly with the custom CA certificate is to initiate a TLS connection to the Admin Router which, after installation, will have a certificate signed by the custom CA. 
+
+In order to do that you need first to obtain the DC/OS CA bundle of the deployed cluster. [This page](/1.10/security/ent/tls-ssl/get-cert/) shows how you can do that.
```
>  you need first to obtain

Is that correct English? I'd say "you first need to obtain" or "first you need to obtain" or "first you need to obtain"

ahairape replied:
right, corrected

sascala commented:
@pavisandhu This page needs a little editing TLC, but is a big improvement

ahairape commented:
@jgehrcke Each of the configuration parameters: ca_certificate_path, ca_certificate_key_path, ca_certificate_chain_path, appears in the old doc page in two places: 
first time in the **table** https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#security-and-authentication, where the parameter name is a link to more detailed description in a following section and there is a brief explanation of the parameter next to that link. Next, there is the dedicated section for each parameter with the detailed explanation.

I could do one of the following:
Option A:
1) leave the table intact
2) replace the text in the corresponding parameter sections https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#ca-certificate-path-enterprise, https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#ca-certificate-key-path-enterprise and https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#ca-certificate-chain-path-enterprise with "See [this page](https://docs.mesosphere.com/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for details"

Option B:
1) For each of the configuration parameters in the table change the link pointing to the specific section in the same document to point to the new documentation page: https://docs.mesosphere.com/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference. Leave the brief description of each parameter intact.
2) Remove the sections https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#ca-certificate-path-enterprise, https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#ca-certificate-key-path-enterprise and https://docs.mesosphere.com/1.10/installing/oss/custom/configuration/configuration-parameters/#ca-certificate-chain-path-enterprise from the old documentation page.

Which option makes more sense to you? Or do you have another suggestion?

ahairape commented:
As discussed today with @jgehrcke, changed the old configuration parameter reference as follows (Option C):
- Removed headings and the text from the config reference
- In the table:
    - in the parameter column removed the link
    - in the description column, put the following sentence for each of the parameters ca_certificate_path, ca_certificate_key_path and ca_certificate_chain_path: Use this to set up a custom CA certificate. See `[this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference)` for a detailed configuration parameter reference

jgehrcke commented:
Super cool, I think we can now ask Thomas to review this for polishing language / style here and there.

ahairape commented on pages/1.10/installing/oss/custom/configuration/configuration-parameters/index.md:
```diff
@@ -74,9 +74,9 @@ This topic provides all available configuration parameters. Except where explici
-| [ca_certificate_path](#ca-certificate-path-enterprise)                   | [enterprise type="inline" size="small" /] Path to a file in the OpenSSL PEM format containing a single X.509 CA certificate. Can be a root (self-issued) certificate or an intermediate (cross-certificate) certificate. |
-| [ca_certificate_key_path](#ca-certificate-key-path-enterprise)           | [enterprise type="inline" size="small" /] Path to a file in the PKCS#8 PEM format containing the private key corresponding to the CA certificate in `ca_certificate_path`. Required if `ca_certificate_path` is specified. |
-| [ca_certificate_chain_path](#ca-certificate-chain-path-enterprise)       | [enterprise type="inline" size="small" /] Path to a file in the OpenSSL PEM format containing the complete CA certification chain required for end-entity certificate verification. Must be left undefined if `ca_certificate_path` is a root CA certificate. Required if `ca_certificate_path` is specified and the specified cer ...
+| ca_certificate_path                   | [enterprise type="inline" size="small" /] Use this to set up a custom CA certificate. See [this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for a detailed configuration parameter reference. |
```
For the reviewer from the doc team:
Please check that the link points correctly to the following page: https://docs.mesosphere.com/1.10/security/ent/tls-ssl/ca-custom/#configuration-parameter-reference

tdk408 replied:
The link seems to work fine.

ahairape commented on pages/1.10/installing/oss/custom/configuration/configuration-parameters/index.md:
```diff
@@ -74,9 +74,9 @@ This topic provides all available configuration parameters. Except where explici
-| [ca_certificate_key_path](#ca-certificate-key-path-enterprise)           | [enterprise type="inline" size="small" /] Path to a file in the PKCS#8 PEM format containing the private key corresponding to the CA certificate in `ca_certificate_path`. Required if `ca_certificate_path` is specified. |
-| [ca_certificate_chain_path](#ca-certificate-chain-path-enterprise)       | [enterprise type="inline" size="small" /] Path to a file in the OpenSSL PEM format containing the complete CA certification chain required for end-entity certificate verification. Must be left undefined if `ca_certificate_path` is a root CA certificate. Required if `ca_certificate_path` is specified and the specified cer ...
+| ca_certificate_path                   | [enterprise type="inline" size="small" /] Use this to set up a custom CA certificate. See [this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for a detailed configuration parameter reference. |
+| ca_certificate_key_path           | [enterprise type="inline" size="small" /] Use this to set up a custom CA certificate. See [this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for a detailed configuration parameter reference. |
```
For the reviewer from the doc team:
Please check that the link points correctly to the following page: https://docs.mesosphere.com/1.10/security/ent/tls-ssl/ca-custom/#configuration-parameter-reference
Thank you!

ahairape commented on pages/1.10/installing/oss/custom/configuration/configuration-parameters/index.md:
```diff
@@ -74,9 +74,9 @@ This topic provides all available configuration parameters. Except where explici
-| [ca_certificate_chain_path](#ca-certificate-chain-path-enterprise)       | [enterprise type="inline" size="small" /] Path to a file in the OpenSSL PEM format containing the complete CA certification chain required for end-entity certificate verification. Must be left undefined if `ca_certificate_path` is a root CA certificate. Required if `ca_certificate_path` is specified and the specified cer ...
+| ca_certificate_path                   | [enterprise type="inline" size="small" /] Use this to set up a custom CA certificate. See [this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for a detailed configuration parameter reference. |
+| ca_certificate_key_path           | [enterprise type="inline" size="small" /] Use this to set up a custom CA certificate. See [this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for a detailed configuration parameter reference. |
+| ca_certificate_chain_path       | [enterprise type="inline" size="small" /] Use this to set up a custom CA certificate. See [this page](/1.10/security/ent/tls-ssl/ca-custom#configuration-parameter-reference) for a detailed configuration parameter reference. |
```
For the reviewer from the doc team:
Please check that the link points correctly to the following page: https://docs.mesosphere.com/1.10/security/ent/tls-ssl/ca-custom/#configuration-parameter-reference
Thank you!

tdk408 replied:
The link on line 79 seems to work.

ahairape commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+
+In order to install DC/OS Enterprise with a custom CA certificate you will need:
+
+- to use the [advanced DC/OS installation method](/1.10/installing/ent/custom/advanced/). Other installation methods are not supported.
```
For reviewer from the doc team:
Please check that the link points correctly to the following page:
https://docs.mesosphere.com/1.10/installing/ent/custom/advanced/
Thank you!

tdk408 replied:
The link seems to be correct.

ahairape commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+## Starting point
+Based on the requirements described above, this is the starting point for the installation:
+
+- The installation of DC/OS Enterprise via the Advanced Installer has been prepared according to the corresponding [documentation](/1.10/installing/ent/custom/advanced/). (up to the section **Install DC/OS** of that documentation)
```
For reviewer from the doc team:
Please check that the link points correctly to the following page:
https://docs.mesosphere.com/1.10/installing/ent/custom/advanced/
Thank you!

tdk408 replied:
The link looks right.

ahairape commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+Note that `ca_certificate_chain_path` must not be present when setting up DC/OS Enterprise with a root certificate as the custom CA certificate.
+
+## Installation
+Proceed with the installation as described in the [documentation of the Advanced Installer](/1.10/installing/ent/custom/advanced/#install-dcos).
```
For reviewer from the doc team:
Please check that the link points correctly to the following page: https://docs.mesosphere.com/1.10/installing/ent/custom/advanced/#install-dcos
Thank you!

tdk408 replied:
The link works.

ahairape commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+## Verify installation
+One method of verifying that the DC/OS Enterprise cluster was installed properly with the custom CA certificate is to initiate a TLS connection to Admin Router which, after installation, will present a certificate signed by the custom CA. 
+
+In order to do that you first need to obtain the DC/OS CA bundle of the deployed cluster. [This page](/1.10/security/ent/tls-ssl/get-cert/) shows how you can do that.
```
For reviewers from the doc team:
Please check that the link points correctly to the following page:
https://docs.mesosphere.com/1.10/security/ent/tls-ssl/get-cert/
Thank you!

tdk408 replied:
the link works.

gpaul commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+
+Furthermore, the file containing the private key `custom_ca.key` corresponding to the custom CA certificate must be owned by the root Unix user and have 0600 permissions set.
+
+If you copy private key file over the network onto the master nodes, the network channel must be adequately protected.
```
s/copy/copy the/

ahairape replied:
fixed

gpaul commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+## ca\_certificate\_path
+Path (relative to the `$DCOS_INSTALL_DIR`) to a file containing a single X.509 CA certificate in the OpenSSL PEM format. For example: `genconf/dcos-ca-certificate.crt`. It is either a *root CA certificate* (“self-signed”) or an *intermediate CA certificate* (“cross-certificate”) signed by some other certificate authority.
+
+If provided, this is the custom CA certificate. It is used as the signing CA certificate, i.e., the DC/OS CA will use this certificate for signing end-entity certificates (the subject of this certificate will be the issuer for certificates signed by the DC/OS CA). If not provided, the DC/OS cluster generates a unique root CA certificate during the initial bootstrap phase and uses that as the sign ...
```
Excellent paragraph

gpaul commented on pages/1.10/security/ent/tls-ssl/ca-custom/index.md:
```diff
@@ -8,26 +8,310 @@ excerpt:
+
+# Configuration parameter reference
+## ca\_certificate\_path
+Path (relative to the `$DCOS_INSTALL_DIR`) to a file containing a single X.509 CA certificate in the OpenSSL PEM format. For example: `genconf/dcos-ca-certificate.crt`. It is either a *root CA certificate* (“self-signed”) or an *intermediate CA certificate* (“cross-certificate”) signed by some other certificate authority.
```
Excellent

ahairape commented:
Thank you very much @jgehrcke, @jongiddy, @mhrabovcin, @tdk408 and @gpaul for your reviews!
@tdk408, could you merge the changes please? (it needs also to be ported to 1.11)

jgehrcke commented:
@ahairape makes sure that this content lands in the docs for 1.10, 1.11, 1.12.
